src/vdaf/mastic.rs

```diff
@@ -379,32 +391,485 @@ where
     }
 }
 
+/// Mastic prepare state
+///
+///  State held by an aggregator between rounds of Mastic. Includes intermediate
+/// state for Szk verification, the output shares currently being validated, and
+/// parameters of Mastic used for encoding.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum MasticPrepareState<F: FieldElement, const SEED_SIZE: usize> {
+    /// Includes state for performing Szk verification at the root.
+    FirstRound(SzkQueryState<SEED_SIZE>, MasticOutputShare<F>, usize, bool),
+    /// In all rounds but the first, SZK is not run, so only the output shares and encoding information are stored.
+    LaterRound(MasticOutputShare<F>, bool),
+}
+
+/// Mastic prepare share
+///
+/// Broadcast message from an aggregator between rounds of Mastic. Includes the
+/// hashed VIDPF proofs for every prefix in the aggregation parameter, and optionally
+/// the verification message for Szk.
+#[derive(Clone, Debug)]
+pub enum MasticPrepareShare<F: FieldElement, const SEED_SIZE: usize> {
+    /// Includes a batched VIDPF proof and an SZK verification message for the root weight.
+    FirstRound(Seed<SEED_SIZE>, SzkQueryShare<F, SEED_SIZE>),
+    /// Includes only a batched VIDPF proof as SZK will not be run.
+    LaterRound(Seed<SEED_SIZE>),
+}
+
+impl<F: FieldElement, const SEED_SIZE: usize> Encode for MasticPrepareShare<F, SEED_SIZE> {
+    fn encode(&self, bytes: &mut Vec<u8>) -> Result<(), CodecError> {
+        match self {
+            MasticPrepareShare::FirstRound(seed, query_share) => {
+                seed.encode(bytes).and_then(|_| query_share.encode(bytes))
+            }
+            MasticPrepareShare::LaterRound(seed) => seed.encode(bytes),
+        }
+    }
+
+    fn encoded_len(&self) -> Option<usize> {
+        match self {
+            MasticPrepareShare::FirstRound(seed, query_share) => {
+                Some(seed.encoded_len()? + query_share.encoded_len()?)
+            }
+            MasticPrepareShare::LaterRound(seed) => seed.encoded_len(),
+        }
+    }
+}
+
+impl<F: FieldElement, const SEED_SIZE: usize> ParameterizedDecode<MasticPrepareState<F, SEED_SIZE>>
+    for MasticPrepareShare<F, SEED_SIZE>
+{
+    fn decode_with_param(
+        prep_state: &MasticPrepareState<F, SEED_SIZE>,
+        bytes: &mut Cursor<&[u8]>,
+    ) -> Result<Self, CodecError> {
+        match prep_state {
+            MasticPrepareState::FirstRound(_, _, verifier_len, requires_joint_rand) => {
+                Ok(MasticPrepareShare::FirstRound(
+                    Seed::<SEED_SIZE>::decode(bytes)?,
```

Comment:
I haven't checked, but the compiler probably can infer this type. Not specifying the type is preferable I think. Here and below.
```suggestion
                    Seed::decode(bytes)?,
```